pre_commit_hooks/destroyed_symlinks.py

```diff
@@ -0,0 +1,108 @@
+PERMS_LINK = b'120000'
+PERMS_NONEXIST = b'000000'
+
+
+def normalize_content(content: bytes) -> bytes:
+    return b'\n'.join(
+        filter(
+            None,
+            map(
+                methodcaller('strip'),
+                content.splitlines(),
+            ),
+        ),
+    )
+
+
+def find_destroyed_symlinks(autofix: bool) -> Sequence[bytes]:
+    destroyed_links = []
+    for line in check_output(
+        ['git', 'status', '--porcelain=v2', '-z'],
+    ).split(b'\0'):
+        splitted = line.split(b' ')
+        if splitted and splitted[0] == ORDINARY_CHANGED_ENTRIES_MARKER:
+            # variable names are taken from
+            # https://git-scm.com/docs/git-status#_changed_tracked_entries
+            _, XY, sub, mH, mI, mW, hH, hI, *path_splitted = splitted
+            path = b' '.join(path_splitted)
+            if all((
+                mH == PERMS_LINK,
+                mI != PERMS_LINK,
+                mI != PERMS_NONEXIST,
+            )):
+                found_destroyed_link = False
+                if hH == hI:
+                    # if old and new hashes are equal, it's not needed to check
+                    # anything more, we've found a destroyed symlink for sure
+                    found_destroyed_link = True
+                else:
+                    # if old and new hashes are *not* equal, it doesn't mean
+                    # that everything is OK - new file may be altered
+                    # by something like trailing-whitespace and/or
+                    # mixed-line-ending hooks so we need to go deeper
+                    index_size = int(
+                        check_output(['git', 'cat-file', '-s', hI]).strip(),
+                    )
+                    # Most filesystems limit path length to 4096 bytes.
+                    # In the worst (insane) case when symlink points to a file
+                    # which path consists of pure newlines and slashes,
+                    # after converting it to Windows line break, its size
+                    # in any case won't be bigger than 4096*2, so if new file
+                    # is bigger than this, we can safely assume that it is not
+                    # destroyed symlink but a valid new file
+                    # instead of the symlink.
+                    if index_size <= 8192:
+                        head_content = normalize_content(
+                            check_output(['git', 'cat-file', '-p', hH]),
+                        )
+                        index_content = normalize_content(
+                            check_output(['git', 'cat-file', '-p', hI]),
+                        )
+                        found_destroyed_link = head_content == index_content
+                if found_destroyed_link:
+                    destroyed_links.append(path)
+                    if autofix:
+                        check_call([
+                            'git',
+                            'update-index',
```

Comment:
pre-commit's stance is to never modify the staging area, I believe this does that so it can't be allowed -- printing this command seems fine though

Reply:
Actually, update-index is redundant here and even harmful — if the file was modified by some other hooks, symlinks would be broken. Instead, I should just unstage the file. Is this allowed?

Reply:
I would just leave it and print a command for the user to run

Reply:
Even if this is opt-in feature (which it is)? May be rename `--autofix` to `--unstage`?

Reply:
there's kinda two problems:
1. pre-commit won't notice the differences
2. it might get auto committed when wrong

if there's a way to modify the file to look like it's the symlink again in the unstaged portion that would be best

Reply:
I've removed `--autofix`, now only the message with an instruction is displayed.

Reply:
> if there's a way to modify the file to look like it's the symlink again in the unstaged portion that would be best

If it was possible, git would have done this. That's a problem.